eng/common/TestResources/Remove-TestResources.ps1

```diff
@@ -213,16 +213,38 @@ $verifyDeleteScript = {
     }
 }
 
+# Get any Key Vaults that will be deleted so they can be purged later if soft delete is enabled.
+$vaults = Get-AzKeyVault -ResourceGroupName "$ResourceGroupName" | ForEach-Object {
+    # Enumerating vaults from a resource group does not return all properties we required.
+    Get-AzKeyVault -VaultName $_.VaultName | Where-Object { $_.EnableSoftDelete }
+}
+
+# You may add additional resource checks that require the resource group to be deleted first before purging here.
+$purgeRequired = !!$vaults
+
 Log "Deleting resource group '$ResourceGroupName'"
-if ($Force) {
+if ($Force -and !$purgeRequired) {
     Remove-AzResourceGroup -Name "$ResourceGroupName" -Force:$Force -AsJob
+    Write-Verbose "Running background job to delete resource group '$ResourceGroupName'"
+
     Retry $verifyDeleteScript 3
-    Write-Verbose "Requested async deletion of resource group '$ResourceGroupName'"
 } else {
     # Don't swallow interactive confirmation when Force is false
     Remove-AzResourceGroup -Name "$ResourceGroupName" -Force:$Force
 }
 
+# Purge any soft deleted vaults since there is now a limit per-subscription.
+foreach ($vault in $vaults) {
+    Log "Attempting to purge Key Vault '$($vault.VaultName)'"
+
+    if ($vault.EnablePurgeProtection) {
+        # We will try anyway but will ignore errors
+        Write-Warning "Key Vault '$($vault.VaultName)' has purge protection enabled and may not be purged for $($vault.SoftDeleteRetentionInDays) days"
```

Comment:
Why is it a "may"? Will this fail in all cases when EnablePurgeProtection returns true or does it just depend on how long it has been? Also is there a way to force override this and delete it anyway?

Reply:
If it was just deleted, yes, this would fail. But if the vault were deleted earlier for some reason than the group, it could work. I figured, "Why not try it?" Errors won't even be added to the `$Error` stream so there's no harm.

Reply:
Is there no way to force delete even for those with purge protection?

Reply:
No. That would defeat the purpose of purge protection.

Reply:
The really unfortunate part is anyone using Key Vaults for customer-managed keys (CMKs) must enable purge protection. I could audit repos' _test-resources.json_ now and see if anyone doing that has at least set the soft delete retention to 7 days. Perhaps we should have the scripts check for that, which we could do in part for #1893.

Reply:
I just pulled the current main from each of the tier 1 lanugage repos (IIRC, azure-sdk-for-cpp doesn't yet use test-resources.json, correct?) and found no one who authored `enablePurgeProtection`. Still, we might want to consider guarding against it. Should I update this PR to handle #1893 as well? I'd have to find KV resources anyway, so enforcing 7 days before purging when purge protection is enabled is incremental.

Reply:
Oops, sorry, there was one. I piped `gi | join-path | select-string` which searched the paths as strings instead of the content. An extra `gi` before `select-string` gives me the file objects I needed.

The only is sdk/search/test-resources.json in azure-sdk-for-net, which I wrote and already sets `softdeleteretentionindays` to 7.